set_bucket_tags/app.py

```diff
@@ -0,0 +1,135 @@
+import json
+import logging
+import traceback
+
+import boto3
+import requests
+
+from crhelper import CfnResource
+
+
+MISSING_BUCKET_NAME_ERROR_MESSAGE = 'BucketName parameter is required'
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+helper = CfnResource(
+  json_logging=False, log_level='DEBUG', boto_level='CRITICAL')
+
+
+def get_s3_client():
+  return boto3.client('s3')
+
+
+def get_bucket_name(event):
+  '''Get the bucket name from event params sent to lambda'''
+  resource_properties = event.get('ResourceProperties')
+  bucket_name = resource_properties.get('BucketName')
+  if not bucket_name:
+    raise ValueError(MISSING_BUCKET_NAME_ERROR_MESSAGE)
+  return bucket_name
+
+
+def get_bucket_tags(bucket_name):
+  '''Look up the bucket tags'''
+  client = get_s3_client()
+  response = client.get_bucket_tagging(Bucket=bucket_name)
+  log.debug(f'S3 bucket tags response: {response}')
+  tags = response.get('TagSet')
+  if not tags or len(tags) == 0:
+    raise Exception(f'No tags returned, received: {response}')
```

Comment:
are existing bucket tags required?  can this work with buckets that do not have existing tags?

Reply:
No, it cannot. This is completely dependent not only on there being tags on the bucket, but on there being a particular tag on the bucket, `aws:servicecatalog:provisioningPrincipalArn`. The `get_principal_id` function will throw an error if that one isn't found.